frame/base_page.py

```
@@ -0,0 +1,152 @@
+
+    def __init__(self, driver, url='', open=False):
+        self.driver = driver
+        self.actions = ActionChains(driver)
+        self.url = BASE_URL + url
+        self.__wait = lambda timeout=TIMEOUT: WebDriverWait(
+            driver, timeout=timeout)
+        if open:
+            self.open()
+
+    @property
+    def locators(self):
+        return [p for p in self.locator.__dict__.items() if p[0].startswith('LOCATOR_')]
+
+    @property
+    def title(self):
+        return self.driver.title
+
+    @property
+    def page_src(self):
+        return self.driver.page_source
+
+    @property
+    def current_url(self):
+        return self.driver.current_url
+
+    def go(self, url):
+        return self.driver.get(url)
+
+    def open(self):
+        return self.driver.get(self.url)
+
+    def at_page(self, title):
+        return self.driver.title == title
+
+    def back(self):
+        return self.driver.back()
+
+    def forward(self):
+        return self.driver.forward()
+
+    def refresh(self):
+        return self.driver.refresh()
+
+    def enter_text(self, locator, text):
+        element = self.find_element(locator)
+        element.click()
+        element.clear()
+        element.send_keys(text)
+        return element
+
+    def enter_text_with_dropdown(self, locator, text):
+        element = self.find_element(locator)
+        element.clear()
+        element.click()
+        element.send_keys(text)
+        try:
+            # handle input with a dropdown
+            self.click(Selector(By.PARTIAL_LINK_TEXT, text))
+        except:
+            element.clear()
+        return element
+
+    def click(self, locator, time=TIMEOUT):
+        try:
+            locator = locator.self
+        except:
+            pass
+        self.__wait(time).until(EC.element_to_be_clickable(locator)).click()
+
+    def hover(self, locator, time=TIMEOUT):
+        try:
+            locator = locator.self
+        except:
+            pass
```

Вот ту тне до конца понятна идея, но каждый раз когда возникает except: pass нужно подумать а точно ли без этой конструкции нельзя обойтись.

---

Да, тут пожалуй корректнее использовать что-то типа
```
if getattr(locator, 'self', None):
    locator = locator.self
```